### PR TITLE
[Misc] Qwen2.5-VL Optimization

### PR DESCRIPTION
**Pull request by wulipc**
Thanks to the VLLM team for their great work! This PR includes the following:

ꔷ Replaced `Qwen2RMSNorm` with vllm's official `RMSNorm`. Extensive testing has confirmed that this change does not introduce any numerical instability.
~~ꔷ Set `_always_apply_prompt_replacements` to `True` by default for Qwen-VL to prevent token replacement errors when inputting multiple images.~~ 
ꔷ Optimized the implementation of `TORCH_SDPA` to avoid creating excessively large attention masks, resulting in a 2-3x speed improvement and a significant reduction in peak video memory usage.

### CODE REVIEW

File: vllm/model_executor/models/qwen2_vl.py [hunk 1]
```diff
@@ -983,6 +990,14 @@ def get_replacement_qwen2vl(item_idx: int, modality: str):
             ) for modality in ("image", "video")
         ]
 
+    def _always_apply_prompt_replacements(self) -> bool:
+        # HF never applies prompt replacements, so we have to do it ourselves
+        # _find_placeholders may incorrectly think that HF has already performed
+        # processing for multi-audio input when the input audios are short
+        # (the corresponding placeholders may take up fewer tokens than
+        # the number of audio items)
+        return True
```
Comment (DarkLight1337) [inline vllm/model_executor/models/qwen2_vl.py:999]: HF does in fact apply prompt replacements. The token mismatch error should be fixed by #13148

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 2]
```diff
@@ -516,8 +506,7 @@ def __init__(
             hidden_size=self.hidden_size,
         )
 
-        # NOTE: We use torch native RMSNorm here for precision purposes.
-        norm_layer = partial(Qwen2RMSNorm, eps=norm_eps)
+        norm_layer = partial(RMSNorm, eps=norm_eps)
```
Comment (ywang96) [inline vllm/model_executor/models/qwen2_5_vl.py:509]: Nice - glad it's confirmed that we can use vLLM `RMSNorm` for this!

File: vllm/model_executor/models/qwen2_5_vl.py [hunk 3]
```diff
@@ -296,20 +302,23 @@ def forward(
                                       "(b s) ... -> b s ...",
                                       b=batch_size)
         elif self.attn_backend == _Backend.TORCH_SDPA:
-            seq_length = q.size(1)
-            q, k, v = (rearrange(x, "b s h d -> b h s d") for x in [q, k, v])
-            attention_mask = torch.zeros([1, seq_length, seq_length],
-                                         device=q.device,
-                                         dtype=torch.bool)
+            # Remove attention_mask for speed & less VRAM.
```
Comment (ywang96) [inline vllm/model_executor/models/qwen2_5_vl.py:305]: Let's change this comment to something like below for less confusion.
```suggestion
            # Execute attention entry by entry for speed & less VRAM.
```

File: vllm/model_executor/models/qwen2_vl.py [hunk 4]
```diff
@@ -336,20 +340,23 @@ def forward(
                                       "(b s) ... -> b s ...",
                                       b=batch_size)
         elif self.attn_backend == _Backend.TORCH_SDPA:
-            seq_length = q.size(1)
-            q, k, v = (rearrange(x, "b s h d -> b h s d") for x in [q, k, v])
-            attention_mask = torch.zeros([1, seq_length, seq_length],
-                                         device=q.device,
-                                         dtype=torch.bool)
+            # Remove attention_mask for speed & less VRAM.
```
Comment (ywang96) [inline vllm/model_executor/models/qwen2_vl.py:343]: same as above

**Review by ywang96 — APPROVED** [review]
Thanks for the PR! Please take a look at @DarkLight1337's comment regarding `_always_apply_prompt_replacements`, otherwise LGTM!

File: vllm/model_executor/models/qwen2_vl.py (hunk 1 above)
Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: Working on it, thanks!

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: During a profile run with a message containing a large number of images (256), we're hitting an OOM error. This is likely due to the transformer preprocessor consuming a significant amount of GPU memory. We've attempted to mitigate this by reducing the preprocessor's `max_pixels` to 100000. This resulted in a different error: `Partial prompt replacement within modality='image' is not supported` which is still occurring now. Interestingly, setting `_always_apply_prompt_replacements` to True bypasses this error. We're unsure if this is a true fix and are actively trying to understand the underlying mechanism. Any information would be greatly appreciated.

It seems the issue stems from the following code block in `_iter_placeholders`:

```
if repl_len_full == 0 or end_idx_full > prompt_len:
	continue
```

Comment (ywang96) [reply vllm/model_executor/models/qwen2_vl.py:999]: > transformer preprocessor consuming a significant amount of GPU memory.

@wulipc This is interesting - are you guys running the preprocessor on GPU? On vLLM this should be running on CPU so I'm surprised.

Also - I assume this is running on V0? Have you tried running it on V1? The profiling space assumption is much more efficient in V1 since we natively support chunked prefill.

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: > @DarkLight1337  During a profile run with a message containing a large number of images (256), we're hitting an OOM error. This is likely due to the transformer preprocessor consuming a significant amount of GPU memory. We've attempted to mitigate this by reducing the preprocessor's `max_pixels` to 100000. This resulted in a different error: `Partial prompt replacement within modality='image' is not supported` which is still occurring now. Interestingly, setting `_always_apply_prompt_replacements` to True bypasses this error. We're unsure if this is a true fix and are actively trying to understand the underlying mechanism. Any information would be greatly appreciated.
 

Can you provide a minimum reproducible example so I can debug it locally?

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: > > transformer preprocessor consuming a significant amount of GPU memory.
> 
> @wulipc This is interesting - are you guys running the preprocessor on GPU? On vLLM this should be running on CPU so I'm surprised.
> 
> Also - I assume this is running on V0? Have you tried running it on V1? The profiling space assumption is much more efficient in V1 since we natively support chunked prefill.

I noticed that the dummy data is truncated to `max_seq_len`, so my previous explanation must have been inaccurate.  Sorry for causing any confusion.

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: @DarkLight1337  I found that the reason for this error `Partial prompt replacement within modality='image' is not supported` is that the number of image tokens is less than the length of the prompt, FYI.

prompt token = 257 = 256 images + 1 video
image token of 100000 pixel  =  121


image

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: If it's because the prompt is truncated, then you might need to further reduce the max image size. I don't think we can handle truncated prompts in general.

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: @DarkLight1337   The truncation I mentioned occurs `after _apply_prompt_replacements`, and `_apply_prompt_replacements` has not been completed when the error is reported;

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: I think this is expected? For example, if

```
len(images) = 3
prompt = <image> * 300
# replacement_texts = [repl_info.get_replacement(item_idx).text for item_idx in len(images)]
replacement_texts = [<image> * 128] * len(images)

then only the first two replacements will be matched, since there aren't sufficient tokens remaining to match the third replacements.
```

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: @DarkLight1337 

In fact, the above error occurs in
```
len(images) = 300
prompt_text = <image> * 300
```

The total number of tokens seems to be wrong.

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: Can you try `disable_mm_preprocessor_cache=True` and see if the problem still appears?

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: > Can you try `disable_mm_preprocessor_cache=True` when initializing `LLM` class and see if the problem still appears?

it still appears @DarkLight1337

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: Hmm, just to check, can you run

```
pytest tests/models/multimodal/processing/test_common.py -k Qwen2-VL
```

(maybe edit the file to set the custom `max_pixels`) and see if you get any error?

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: @DarkLight1337  Yes, it still appears,  and you can reproduce it with the following configuration：

1. change `max_pixels`  in `preprocessor_config.json`  to 100000;
2. change the image number to 256 (3 -> 256); 
https://github.com/vllm-project/vllm/blob/cb944d5818398281d5e777824ee394d5455d8b8b/tests/models/multimodal/processing/test_common.py#L56

image

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: Ok I think I found the problem. The number of feature tokens per image (e.g. 49 for a 195x192 image) is smaller than the number of images (in this case 256), so it becomes difficult to distinguish whether HF processor has been applied or not.

Comment (DarkLight1337) [reply vllm/model_executor/models/qwen2_vl.py:999]: I have drafted a fix: https://github.com/vllm-project/vllm/pull/13215, can you try it out?

Comment (wulipc) [reply vllm/model_executor/models/qwen2_vl.py:999]: > I have drafted a fix: #13215, can you try it out?

It's working fine now, thanks for the fix!

**Comment by wulipc** [comment]
@DarkLight1337  @ywang96  I think this PR can be merged into the main, and I have deleted the `_always_apply_prompt_replacements`  func and merged the latest main code.

File: vllm/model_executor/models/qwen2_5_vl.py (hunk 3 above)
Comment (wulipc) [reply vllm/model_executor/models/qwen2_5_vl.py]: > Can you resolve these comments as well?

@DarkLight1337  done!
